subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java

```diff
@@ -0,0 +1,133 @@
+import org.gradle.api.internal.project.HoldsProjectState;
+import org.gradle.api.internal.project.ProjectState;
+import org.gradle.api.internal.project.ProjectStateRegistry;
+import org.gradle.internal.Describables;
+import org.gradle.internal.Factory;
+import org.gradle.internal.build.CompositeBuildParticipantBuildState;
+import org.gradle.internal.build.IncludedBuildState;
+import org.gradle.internal.component.local.model.LocalComponentGraphResolveState;
+import org.gradle.internal.component.local.model.LocalComponentGraphResolveStateFactory;
+import org.gradle.internal.component.local.model.LocalComponentMetadata;
+import org.gradle.internal.model.CalculatedValueContainer;
+import org.gradle.internal.model.CalculatedValueContainerFactory;
+
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Default implementation of {@link BuildTreeLocalComponentProvider}.
+ */
+public class DefaultBuildTreeLocalComponentProvider implements BuildTreeLocalComponentProvider, HoldsProjectState {
+
+    private final ProjectStateRegistry projectStateRegistry;
+    private final LocalComponentGraphResolveStateFactory resolveStateFactory;
+    private final LocalComponentCache localComponentCache;
+    private final LocalComponentProvider localComponentProvider;
+
+    private final ConcurrentCache localProjects;
+    private final ConcurrentCache foreignProjects;
+
+    public DefaultBuildTreeLocalComponentProvider(
+        ProjectStateRegistry projectStateRegistry,
+        CalculatedValueContainerFactory calculatedValueContainerFactory,
+        LocalComponentGraphResolveStateFactory resolveStateFactory,
+        LocalComponentCache localComponentCache,
+        LocalComponentProvider localComponentProvider
+    ) {
+        this.projectStateRegistry = projectStateRegistry;
+        this.resolveStateFactory = resolveStateFactory;
+        this.localComponentCache = localComponentCache;
+        this.localComponentProvider = localComponentProvider;
+
+        this.localProjects = new ConcurrentCache(calculatedValueContainerFactory);
+        this.foreignProjects = new ConcurrentCache(calculatedValueContainerFactory);
+    }
+
+    @Override
+    public LocalComponentGraphResolveState getLocalComponent(ProjectComponentIdentifier projectIdentifier) {
+        return getLocalComponent(projectIdentifier, projectStateRegistry.stateFor(projectIdentifier));
+    }
+
+    private LocalComponentGraphResolveState getLocalComponent(ProjectComponentIdentifier projectIdentifier, ProjectState projectState) {
+        return localProjects.computeIfAbsent(projectIdentifier, () -> localComponentCache.computeIfAbsent(projectState, localComponentProvider::getComponent));
+    }
+
+    @Override
+    public LocalComponentGraphResolveState getForeignComponent(ProjectComponentIdentifier projectIdentifier) {
+        return foreignProjects.computeIfAbsent(projectIdentifier, () -> getComponentInAnotherBuild(projectIdentifier));
```

Comment:
💭 `getComponentInAnotherBuild` is clearer than `getForeignComponent` to me.  Foreign makes me think at the level of components of external dependencies, not the level of projects within a build.

Reply:
I was just re-using the terminology set elsewhere. [Foreign build identifier](https://github.com/gradle/gradle/blob/ce40262847aa87f2b53965b4b1e9bc9d8fd89ee4/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ForeignBuildIdentifier.java#L24) is an established name used for identifiers of builds from a different build than the current one. 

I think a name like `getLocalComponentWithForeighId` would be a fitting name here